core/include/vecmem/utils/debug.hpp

```diff
@@ -0,0 +1,94 @@
+/** VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+#pragma once
+
+/// Function name to use for printout operations
+#ifndef VECMEM_PRINTF
+#   if defined(SYCL_LANGUAGE_VERSION) || defined(CL_SYCL_LANGUAGE_VERSION)
+#      define VECMEM_PRINTF cl::sycl::ONEAPI::experimental::printf
+#   else
+#      define VECMEM_PRINTF printf
+#   endif
+#endif // not VECMEM_PRINTF
+
+// Attribute(s) to use on the printed message character string variables
+#ifndef VECMEM_MSG_ATTRIBUTES
+#   ifdef __SYCL_DEVICE_ONLY__
+#      define VECMEM_MSG_ATTRIBUTES __attribute__((opencl_constant))
+#   else
+#      define VECMEM_MSG_ATTRIBUTES
+#   endif
+#endif // not VECMEM_MSG_ATTRIBUTES
+
+/// Set a default maximum level for the printed messages
+#ifndef VECMEM_DEBUG_MSG_LVL
+#   define VECMEM_DEBUG_MSG_LVL 0
+#endif // not VECMEM_DEBUG_MSG_LVL
+
+/// Helper macro for using the underlying @c printf function
+///
+/// This intermediate function is really only necessary for SYCL compatibility.
+/// Since one needs to very carefully declare the character constant that would
+/// be given to the oneAPI printf function.
+///
+#define __VECMEM_PRINT_MSG(MSG, ...)                                           \
+   do {                                                                        \
+      const VECMEM_MSG_ATTRIBUTES char __msg[] = MSG;                          \
+      VECMEM_PRINTF( __msg, __VA_ARGS__ );                                     \
+   } while( false )
+
+/// Print macro for "level 1" debug messages
+#if VECMEM_DEBUG_MSG_LVL >= 1
+#   define __VECMEM_PRINT_1(MSG, ...)                                          \
+           __VECMEM_PRINT_MSG(MSG, __VA_ARGS__ )
+#else
+#   define __VECMEM_PRINT_1(MSG, ...)
+#endif
+
+/// Print macro for "level 2" debug messages
+#if VECMEM_DEBUG_MSG_LVL >= 2
+#   define __VECMEM_PRINT_2(MSG, ...)                                          \
+           __VECMEM_PRINT_MSG(MSG, __VA_ARGS__ )
+#else
+#   define __VECMEM_PRINT_2(MSG, ...)
+#endif
+
+/// Print macro for "level 3" debug messages
+#if VECMEM_DEBUG_MSG_LVL >= 3
+#   define __VECMEM_PRINT_3(MSG, ...)                                          \
+           __VECMEM_PRINT_MSG(MSG, __VA_ARGS__ )
+#else
+#   define __VECMEM_PRINT_3(MSG, ...)
+#endif
+
+/// Print macro for "level 4" debug messages
+#if VECMEM_DEBUG_MSG_LVL >= 4
+#   define __VECMEM_PRINT_4(MSG, ...)                                          \
+           __VECMEM_PRINT_MSG(MSG, __VA_ARGS__ )
+#else
+#   define __VECMEM_PRINT_4(MSG, ...)
+#endif
+
+/// Print macro for "level 5" debug messages
+#if VECMEM_DEBUG_MSG_LVL >= 5
+#   define __VECMEM_PRINT_5(MSG, ...)                                          \
+           __VECMEM_PRINT_MSG(MSG, __VA_ARGS__ )
+#else
+#   define __VECMEM_PRINT_5(MSG, ...)
+#endif
```

Comment:
If you know how to do this with less number of lines, I'm all ears. I couldn't come up with anything better. :frowning: